test/infrastructure/container/fake.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"context"
+	"io"
+)
+
+var runContainerCallLog []RunContainerArgs
+var deleteContainerCallLog []string
+var killContainerCallLog []KillContainerArgs
+var execContainerCallLog []ExecContainerArgs
+
+// RunContainerArgs contains the arguments passed to calls to RunContainer.
+type RunContainerArgs struct {
+	RunConfig *RunContainerInput
+	Output    io.Writer
+}
+
+// KillContainerArgs contains the arguments passed to calls to Kill.
+type KillContainerArgs struct {
+	Container string
+	Signal    string
+}
+
+// ExecContainerArgs contains the arguments passed to calls to ExecContainer.
+type ExecContainerArgs struct {
+	ContainerName string
+	Config        *ExecContainerInput
+	Command       string
+	Args          []string
+}
+
+type FakeRuntime struct {
+}
+
+// NewFakeClient gets a client for testing.
+func NewFakeClient() (Runtime, error) {
+	return &FakeRuntime{}, nil
+}
+
+// FakeClientContext gets a context to use configured with our testing fake.
+func (f *FakeRuntime) FakeClientContext() context.Context {
+	return context.WithValue(context.Background(), providerKey{}, *f)
+}
+
+// SaveContainerImage saves a Docker image to the file specified by dest.
+func (f *FakeRuntime) SaveContainerImage(ctx context.Context, image, dest string) error {
+	return nil
+}
+
+// PullContainerImageIfNotExists triggers the Docker engine to pull an image, but only if it doesn't
+// already exist. This is important when we're using locally built images in CI which
+// do not exist remotely.
+func (f *FakeRuntime) PullContainerImageIfNotExists(ctx context.Context, image string) error {
+	return nil
+}
+
+// GetHostPort looks up the host port bound for the port and protocol (e.g. "6443/tcp").
+func (f *FakeRuntime) GetHostPort(ctx context.Context, containerName, portAndProtocol string) (string, error) {
+	return "", nil
+}
+
+// ExecContainer executes a command in a running container and writes any output to the provided writer.
+func (f *FakeRuntime) ExecContainer(ctxd context.Context, containerName string, config *ExecContainerInput, command string, args ...string) error {
+	execContainerCallLog = append(execContainerCallLog, ExecContainerArgs{
+		ContainerName: containerName,
+		Config:        config,
+		Command:       command,
+		Args:          args,
+	})
+	return nil
+}
+
+func (f *FakeRuntime) ExecContainerCalls() []ExecContainerArgs {
```

Comment:
Please add go doc to public methods (not sure why this is not cached by linters)

Reply:
I noticed we do not enforce that check in our linter configuration. It's especially visible to me because my local IDE linting settings do flag it, even though we are not enforcing it.

I will get godocs added to all new methods. We may want to follow up project-wide and enable that in our linter config to make sure it gets enforced everywhere.

Reply:
Enabling this check for all of `infrastructure/docker` here so the linter will flag things like this: https://github.com/kubernetes-sigs/cluster-api/pull/5817